docs/proposal/kcl-checksum-verification-checker-module-design.md

```diff
@@ -0,0 +1,139 @@
+# Checker module design for kcl dependency
+```Go
+package Checker
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/hashicorp/go-version"
+	pkg "kcl-lang.io/kpm/pkg/package"
+)
+
+// Checker is the interface for all dependency checkers.
+type Checker interface {
+	Check(d pkg.Dependency, localCachePath string) error
+}
+
+type DepChecker struct {
+	checkers []Checker
+}
+
+func NewDepChecker(checkers ...Checker) *DepChecker {
+	return &DepChecker{checkers: checkers}
+}
+
+// Check runs all the individual checks for a dependency.
+func (c *DepChecker) Check(d pkg.Dependency, localCachePath string) error {
+	for _, checker := range c.checkers {
+		if err := checker.Check(d, localCachePath); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// IdentChecker validates the dependency name.
+type IdentChecker struct{}
+
+func (c *IdentChecker) Check(d pkg.Dependency, localCachePath string) error {
+	if !isValidDependencyName(d.Name) {
+		return fmt.Errorf("invalid dependency name: %s", d.Name)
+	}
+	return nil
+}
+
+// VersionChecker validates the dependency version.
+type VersionChecker struct{}
+
+func (c *VersionChecker) Check(d pkg.Dependency, localCachePath string) error {
+	if !isValidDependencyVersion(d.Version) {
+		return fmt.Errorf("invalid dependency version: %s for %s", d.Version, d.Name)
+	}
+	return nil
+}
+
+// SumChecker validates the dependency checksum.
+type SumChecker struct{}
+
+func (c *SumChecker) Check(d pkg.Dependency, localCachePath string) error {
+	gotSum, err := HashLocalCache(localCachePath)
+	if err != nil {
+		return fmt.Errorf("failed to calculate checksum: %w", err)
+	}
+	if d.Sum != gotSum {
```

Comment:
Is it compared with the sum value calculated by the local file ? I remember it was to be compared with the sum stored remotely or sum local cache. The purpose of checking the checksum is to prevent security risks caused by malicious replacement of remote third-party dependencies.

Reply:
Thank you for the feedback. Yes, the method calculates the checksum of the local file or cache. 
The current implementation ensures that the local file matches the expected checksum stored in the dependency object (`d.Sum`). This object should ideally be populated with the checksum value retrieved from a trusted remote source or sum local cache to ensure we are validating the file against a reliable source.

Reply:
Are you suggesting that `d.Sum` may be compromised, and that we should include the actual design for retrieving the checksum from the mod file or remotely?

Reply:
Hi @NishantBansal2003 😄 

I don't understand what you want to express. My understanding is that the main function of this checksum is to determine whether the sum in the lock file is the same as the sum on the remote end, ensuring the security of the package. If you think there might be a different reason or a different usage, you might need to provide things like rust or go that they do with checksum.

Reply:
Apologies for the previous approach. Earlier, I thought that `d.Sum` in KPM stores the trusted checksum for a dependency. However, after carefully reviewing the KPM code, I now understand that `d.Sum` retrieves the checksum from the OCI registry or calculates it using `HashDir` (which is to be stored in the lock file). This checksum needs to be compared with a remote trusted source to ensure the integrity of the dependency. Am I correct now?
https://github.com/kcl-lang/kpm/blob/bea45753c66734d7f5ebe2a44fe544d14402db13/pkg/client/client.go#L1226-L1236

Reply:
Yes

Reply:
Sure, now I think I have some clarity. I will update the design doc based on the reviews.